lib/src/Uncertainty/Algorithm/WeightedExperiments/TensorProductExperiment.cxx

```diff
@@ -0,0 +1,187 @@
+/**
+ *  @brief Abstract top-level view of a TensorProductExperiment
+ *
+ *  Copyright 2005-2021 Airbus-EDF-IMACS-ONERA-Phimeca
+ *
+ *  This library is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU Lesser General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This library is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Lesser General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Lesser General Public License
+ *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+#include "openturns/TensorProductExperiment.hxx"
+#include "openturns/Tuples.hxx"
+#include "openturns/Exception.hxx"
+#include "openturns/PersistentObjectFactory.hxx"
+#include "openturns/Log.hxx"
+#include "openturns/BlockIndependentDistribution.hxx"
+#include <iostream>
+
+BEGIN_NAMESPACE_OPENTURNS
+
+CLASSNAMEINIT(TensorProductExperiment)
+
+typedef Collection< Point > PointCollection;
+
+/* Default constructor */
+TensorProductExperiment::TensorProductExperiment()
+  : WeightedExperimentImplementation()
+  , collection_(0)
+  , nodes_(0, 0)
+{
+  // Nothing to do
+}
+
+/* Constructor with parameters */
+TensorProductExperiment::TensorProductExperiment(const WeightedExperimentCollection & collection)
+  : WeightedExperimentImplementation()
+  , collection_(collection)
+  , nodes_(0, 0)
+{
+  const UnsignedInteger numberOfMarginalExperiments = collection_.getSize();
+  BlockIndependentDistribution::DistributionCollection distributionCollection(numberOfMarginalExperiments);
+  UnsignedInteger size = 1;
+  for (UnsignedInteger i = 0; i < numberOfMarginalExperiments; ++i)
+  {
+    size *= collection_[i].getSize();
+    distributionCollection[i] = collection_[i].getDistribution();
+  }
+  WeightedExperimentImplementation::setSize(size);
+  const BlockIndependentDistribution distribution(distributionCollection);
+  WeightedExperimentImplementation::setDistribution(distribution);
+}
+
+/* Virtual constructor */
+TensorProductExperiment * TensorProductExperiment::clone() const
+{
+  return new TensorProductExperiment(*this);
+}
+
+/* String converter */
+String TensorProductExperiment::__repr__() const
+{
+  OSS oss;
+  oss << "class=" << GetClassName()
+    << " name=" << getName()
+    << " collection=" << collection_;
+  return oss;
+}
+
+Bool TensorProductExperiment::hasUniformWeights() const
+{
+  const UnsignedInteger numberOfMarginalExperiments = collection_.getSize();
+  bool hasUniformWeights = true;
+  for (UnsignedInteger i = 0; i < numberOfMarginalExperiments; ++i)
+  {
+    if (!collection_[i].hasUniformWeights())
+    {
+      hasUniformWeights = false;
+      break;
+    }
+  }
+
+  return hasUniformWeights;
+}
```

Comment:
Another way:
```
{
  for (UnsignedInteger i = 0; i < collection_.getSize(); ++i)
    if (!collection_[i].hasUniformWeights()) return false;
  return true;
}
```
Shorter to read and not so obfuscated IMO

Reply:
@mbaudin47 It is not because it is written in stackoverflow that it is the Great Truth ;-)! I prefer to have a compact 3 lines implementation with an early exit rather than a 12 lines implementation with an intermediate variable and additional flow control just to have a single return. It is of course another story for a 100+ lines method!

Still +1 even with a bloated implementation :)

Reply:
Oups. Actually, it seems that this commit was not pushed so that the merge does not have the corresponding update. I move it to another PR. Sorry for the mess.